sympy/solvers/ode/systems.py

```diff
@@ -486,6 +446,79 @@ def _canonical_equations(eqs, funcs, t):
     return canon_eqs
 
 
+def _is_commutative_anti_derivative(A, t):
+    B = integrate(A, t)
+    is_commuting = (B*A - A*B).applyfunc(expand).applyfunc(factor_terms).is_zero_matrix
+
+    return B, is_commuting
+
+
+def _neq_linear_first_order_nonconst_coeff_homogeneous(match_):
+    r"""
+    System of n first-order nonconstant-coefficient linear homogeneous differential equations
+
+    .. math::
+        X' = A(t) X
+
+    where $X$ is the vector of $n$ dependent variables, $t$ is the dependent variable, $X'$
+    is the first order differential of $X$ with respect to $t$ and $A(t)$ is a $n \times n$
+    coefficient matrix.
+
+    Let us define $B$ as antiderivative of coefficient matrix $A$:
+
+    .. math::
+        B(t) = \int A(t) dt
+
+    If the system of ODEs defined above is such that its antiderivative $B(t)$ commutes with
+    $A(t)$ itself, then, the solution of the above system is given as:
+
+    .. math::
+        X = \exp(B(t)) C
+
+    where $C$ is the vector of constants.
+
+    """
+
+    # Some parts of code is repeated, this needs to be taken care of
+    # The constant vector obtained here can be done so in the match
+    # function itself.
+    eq = match_['eq']
+    func = match_['func']
+    fc = match_['func_coeff']
+    n = len(eq)
+    t = list(list(eq[0].atoms(Derivative))[0].atoms(Symbol))[0]
+    constants = numbered_symbols(prefix='C', cls=Symbol, start=1)
+
+    # This needs to be modified in future so that fc is only of type Matrix
+    M = -fc if type(fc) is Matrix else Matrix(n, n, lambda i,j:-fc[i,func[j],0])
+
+    Cvect = Matrix(list(next(constants) for _ in range(n)))
+
+    # The code in if block will be removed when it is made sure
+    # that the code works without the statements in if block.
+    if "commutative_antiderivative" not in match_:
+        B, is_commuting = _is_commutative_anti_derivative(M, t)
+
+        # This course is subject to change
+        if not is_commuting:
+            return None
+
+    else:
+        B = match_['commutative_antiderivative']
+
+    sol_vector = B.exp() * Cvect
+
+    # 19185: This isn't properly collecting the expressions, we can do something like
+    # collect(expand(s), ordered(.... where we expand the s below so that we can collect
+    # the powers where it isn't possible.
+    # For example: collect(a*(x+2), x) == a*(x + 2) but
+    # collect(expand(a*(x+2)), x) == a*x + 2*a
+    sol_vector = [collect(s, ordered(s.atoms(exp)), exact=True) for s in sol_vector]
```

Comment:
@oscarbenjamin As mentioned in this marker, some of the expressions are not being collected properly. Can we use ```expand``` function here as a workaround?

Reply:
Can you give an example of what `sol_vector` is here?

Reply:
For ```dsolve(eqs)``` where ```eqs = [Eq(f(x).diff(x), x * f(x) + g(x)), Eq(g(x).diff(x), -f(x) + x * g(x))]```, 

```
sol_vector before collect is Matrix([[C1*(exp(x**2/2 - I*x)/2 + exp(x**2/2 + I*x)/2)
 + C2*(I*exp(x**2/2 - I*x)/2 - I*exp(x**2/2 + I*x)/2)], 
[C1*(-I*exp(x**2/2 - I*x)/2 + I*exp(x**2/2 + I*x)/2) 
+ C2*(exp(x**2/2 - I*x)/2 + exp(x**2/2 + I*x)/2)]])

sol_vector after collect is [C1*(exp(x**2/2 - I*x)/2 + exp(x**2/2 + I*x)/2)
+ C2*(I*exp(x**2/2 - I*x)/2 - I*exp(x**2/2 + I*x)/2),
C1*(-I*exp(x**2/2 - I*x)/2 + I*exp(x**2/2 + I*x)/2) + C2*(exp(x**2/2 - I*x)/2 + exp(x**2/2 + I*x)/2)]
```

Reply:
We could use `expand_mul` I guess. In this example ideally the solution would come out like this:
```julia
In [31]: M=ImmutableMatrix([[C1*(exp(x**2/2 - I*x)/2 + exp(x**2/2 + I*x)/2) 
    ...:  + C2*(I*exp(x**2/2 - I*x)/2 - I*exp(x**2/2 + I*x)/2)],  
    ...: [C1*(-I*exp(x**2/2 - I*x)/2 + I*exp(x**2/2 + I*x)/2)  
    ...: + C2*(exp(x**2/2 - I*x)/2 + exp(x**2/2 + I*x)/2)]])                                                                                   

In [32]: (M / exp(x**2/2)).expand().rewrite(cos).expand() * exp(x**2/2)                                                                        
Out[32]: 
⎡                          2 ⎤
⎢                         x  ⎥
⎢                         ── ⎥
⎢                         2  ⎥
⎢(C₁⋅cos(x) + C₂⋅sin(x))⋅ℯ   ⎥
⎢                            ⎥
⎢                           2⎥
⎢                          x ⎥
⎢                          ──⎥
⎢                          2 ⎥
⎣(-C₁⋅sin(x) + C₂⋅cos(x))⋅ℯ  ⎦
```

Reply:
Another possibility is that maybe we should make `x` real:
```julia
In [115]: x = Symbol('x', real=True)                                                                                                           

In [116]: B = Matrix([[-x**2/2, -x], [x, -x**2/2]])                                                                                            

In [117]: B.exp()                                                                                                                              
Out[117]: 
⎡   2              2        ⎤
⎢ -x             -x         ⎥
⎢ ────           ────       ⎥
⎢  2              2         ⎥
⎢ℯ    ⋅cos(x)  -ℯ    ⋅sin(x)⎥
⎢                           ⎥
⎢   2             2         ⎥
⎢ -x            -x          ⎥
⎢ ────          ────        ⎥
⎢  2             2          ⎥
⎣ℯ    ⋅sin(x)  ℯ    ⋅cos(x) ⎦
```

Reply:
> We could use expand_mul I guess. In this example ideally the solution would come out like this:

@oscarbenjamin Yes, ```expand_mul``` would be better since that's what we require for collecting expressions that are not being collected due to the brackets.

Reply:
> Another possibility is that maybe we should make x real

I would add a test case regarding this.